src/proper_fsm.erl

```diff
@@ -314,6 +367,108 @@ postcondition(#state{name = From, data = Data, mod = Mod}, Call, Res) ->
+  State = #state{name = Name, data = Data, mod = Mod},
+  ?LET([_ | Cmds], proper_statem:commands(?MODULE, State),
+       {finalize_weights(State, Cmds, maps:new()),
+        [{init, InitialState} | Cmds]}).
+
+-spec next_commands_gen(mod_name()) -> next_fun().
+next_commands_gen(Mod) ->
+  fun ({Weights, _Cmds}, {_D, T}) ->
+      State = initial_state(Mod),
+      NewWeights = proper_statem:next_weights(Weights, T),
+      CmdsGen = ?LET([_ | Cmds],
+                     proper_statem:next_gen(?MODULE, NewWeights, State),
+                     Cmds),
+      ?SHRINK(
+         ?LET(Cmds, ?LAZY(CmdsGen),
+              {finalize_weights(State, Cmds, NewWeights), Cmds}),
+         [CmdsGen])
+  end.
+
+-spec next_commands_gen(mod_name(), fsm_state()) -> next_fun().
+next_commands_gen(Mod, {Name, Data} = InitialState) ->
+  fun ({Weights, _Cmds}, {_D, T}) ->
+      State = #state{name = Name, data = Data, mod = Mod},
+      NewWeights = proper_statem:next_weights(Weights, T),
+      CmdsGen = ?LET([_ | Cmds],
+                     proper_statem:next_gen(?MODULE, NewWeights, State),
+                     [{init, InitialState} | Cmds]),
+      ?SHRINK(
+         ?LET(Cmds, ?LAZY(CmdsGen),
+              {finalize_weights(State, Cmds, NewWeights), Cmds}),
+         [CmdsGen])
+  end.
+
+%% Commands are selected according to the weights found in the
+%% map provided, or from a `weight/3' callback.
+
+%% @private
+-spec select_command(state(), weights()) -> proper_types:type().
+select_command(State, Weights) ->
+  #state{name = From, data = Data, mod = Mod} = State,
+  SelectAux =
+    fun ({To, {call, _Mod, Call, _Args} = SymbCall}) ->
+        RealTo = cook_history(From, To),
+        Key = {From, RealTo, Call},
+        case maps:is_key(Key, Weights) of
+          true -> {maps:get(Key, Weights), SymbCall};
+          false ->
+            case is_exported(Mod, {weight, 3}) of
+              true -> {Mod:weight(From, RealTo, SymbCall), SymbCall};
+              false -> {1, SymbCall}
+            end
+        end
+    end,
+  ?LET(Transitions,
+       ?LAZY(get_transitions(Mod, From, Data)),
+       begin
+         ValidTransitions = [{To, Cmd} || {To, Cmd} <- Transitions,
+                                          precondition(State, Cmd)],
+         WeightedCmds = lists:map(SelectAux, ValidTransitions),
+         proper_types:frequency(WeightedCmds)
+       end).
+
+%% Track all the state transitions and add them to the weights map.
+
+-spec finalize_weights(state(), command_list(), weights()) -> weights().
+finalize_weights(_S, [], Weights) -> Weights;
+finalize_weights(InitialState, Cmds, Weights) ->
+  FinWeightsAux =
+    fun ({init, _InitialState}, Acc) -> Acc;
+        ({set, Var, {call, _Mod, Call, _Args} = SymbCall}, {State, Ws}) ->
+        NextState = next_state(State, Var, SymbCall),
+        From = State#state.name,
+        To = NextState#state.name,
+        Key = {From, To, Call},
+        case maps:is_key(Key, Ws) of
```

Comment:
Similar comment to the above here about `NewWs` computation and with `{NextState, NewWs}` immediately below it.